core/src/main/java/smile/classification/IsotonicRegressionScaling.java

```diff
@@ -0,0 +1,220 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+package smile.classification;
+
+import java.io.Serializable;
+import java.util.Arrays;
+
+import smile.sort.QuickSort;
+
+/**
+ * An alternative approach to probability calibration is to fit an isotonic
+ * regression model to an ill-calibrated probability model. This has been shown
+ * to work better than Platt scaling, in particular when enough training data is
+ * available.
+ * 
+ * <h2>References</h2>
+ * <ol>
+ * <li>Niculescu-Mizil, Alexandru; Caruana, Rich: Predicting Good Probabilities
+ * With Supervised Learning, 2005.</li>
+ * </ol>
+ *
+ * @author rayeaster
+ */
+public class IsotonicRegressionScaling implements Serializable {
+    private static final long serialVersionUID = 1L;
+    
+    private double trainMixScore = Double.NaN;
+    private double trainMaxScore = Double.NaN;
+
+    /** setep-wise constant functions */
+    private static class StepwiseConsFunc {
+        double lowScore;
+        double highScore;
+        double val;
+        int weight = 1;
+
+        StepwiseConsFunc next;
+        StepwiseConsFunc previous;
+
+        StepwiseConsFunc(double lowScore, double highScore, double val) {
+            this.lowScore = lowScore;
+            this.highScore = highScore;
+            this.val = val;
+        }
+
+        void setNext(StepwiseConsFunc next) {
+            this.next = next;
+        }
+
+        void setPrevious(StepwiseConsFunc previous) {
+            this.previous = previous;
+        }
+
+        public int getWeight() {
+            return weight;
+        }
+
+        public void setWeight(int weight) {
+            this.weight = weight;
+        }
+
+        boolean matchValue(double score) {
+            if ((score - lowScore) > 0.0 && (highScore - score) >= 0.0) {
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        double getVal() {
+            return this.val;
+        }
+    }
+
+    private StepwiseConsFunc startConstantFuncs;
+
+    /**
+     * Trains the Isotonic Regression scaling.
+     * 
+     * @param scores
+     *            The predicted scores.
+     * @param y
+     *            The training labels.
+     */
+    public IsotonicRegressionScaling(double[] scores, int[] y) {
+        initStepwiseConstFuncs(scores, y);
+
+        StepwiseConsFunc current = startConstantFuncs.next;
+        StepwiseConsFunc last = startConstantFuncs;
+        
+        while (current != null) {
```

Comment:
We need actually check the PAV in multiple iterations. Think about it as an exercise.